collector/mssql.go

```diff
@@ -3731,6 +3833,123 @@ func (c *MSSQLCollector) collectSQLStats(ctx *ScrapeContext, ch chan<- prometheu
 	return nil, nil
 }
 
+// Win32_PerfRawData_MSSQLSERVER_SQLServerWaitStatistics docs:
+// - https://docs.microsoft.com/en-us/sql/relational-databases/performance-monitor/sql-server-wait-statistics-object
+type mssqlWaitStatistics struct {
+	Name                                   string
+	WaitStatsLockWaits                     float64 `perflib:"Lock waits"`
+	WaitStatsMemoryGrantQueueWaits         float64 `perflib:"Memory grant queue waits"`
+	WaitStatsThreadSafeMemoryObjectsWaits  float64 `perflib:"Thread-safe memory objects waits"`
+	WaitStatsLogWriteWaits                 float64 `perflib:"Log write waits"`
+	WaitStatsLogBufferWaits                float64 `perflib:"Log buffer waits"`
+	WaitStatsNetworkIOWaits                float64 `perflib:"Network IO waits"`
+	WaitStatsPageIOLatchWaits              float64 `perflib:"Page IO latch waits"`
+	WaitStatsPageLatchWaits                float64 `perflib:"Page latch waits"`
+	WaitStatsNonpageLatchWaits             float64 `perflib:"Non-Page latch waits"`
+	WaitStatsWaitForTheWorkerWaits         float64 `perflib:"Wait for the worker"`
+	WaitStatsWorkspaceSynchronizationWaits float64 `perflib:"Workspace synchronization waits"`
+	WaitStatsTransactionOwnershipWaits     float64 `perflib:"Transaction ownership waits"`
+}
+
+func (c *MSSQLCollector) collectWaitStats(ctx *ScrapeContext, ch chan<- prometheus.Metric, sqlInstance string) (*prometheus.Desc, error) {
+	var dst []mssqlWaitStatistics
+	log.Debugf("mssql_waitstats collector iterating sql instance %s.", sqlInstance)
+
+	if err := unmarshalObject(ctx.perfObjects[mssqlGetPerfObjectName(sqlInstance, "waitstats")], &dst); err != nil {
+		return nil, err
+	}
+
+	for _, v := range dst {
+		item := v.Name
+
+		ch <- prometheus.MustNewConstMetric(
+			c.WaitStatsLockWaits,
+			prometheus.GaugeValue,
```

Comment:
This should be `CounterValue`, right? (That's what you specify in the docs)

Reply:
Thanks for the feedback, the code was correct (according to the sql docs), I had to update my docs.

Reply:
Huh, that's real unexpected... Does it report current locks being waited on?

Reply:
I'm relatively new to SQL server statistics, so I might be wrong, but according to [this doc](https://docs.microsoft.com/en-us/sql/relational-databases/performance-monitor/sql-server-wait-statistics-object?view=sql-server-2017)

* the Wait Stats object contains a bunch of counters (Lock Waits, Wait for the Worker, etc.), and
* each counter contains the following instances: 
  * "Average wait time (ms)"
  * "Cumulative wait time (ms) per second"
  * "Waits in progress" (which is just the number of processes currently waiting)
  * "Waits started per second"

Since these metrics can go up or down, it seems like GaugeValue would be the correct choice.

Reply:
Actually, I suppose that "Cumulative wait time" can only go up and so could be a Prometheus counter, but given how the other collector functions loop over the objects (for multi-object stats, like [LockWaits](https://github.com/prometheus-community/windows_exporter/blob/19fee044bf2ec26e1e698d3faf4a3cf8cf33d026/collector/mssql.go#L3377), it seemed more reasonable to keep this as a gauge rather than try to carve out a special handler for this one case.

Also, just to make sure I'm thinking about this in the right way, in the code for DatabaseDataFilesSizeKB ([here](https://github.com/prometheus-community/windows_exporter/blob/19fee044bf2ec26e1e698d3faf4a3cf8cf33d026/collector/mssql.go#L2856-L2861)), we're emitting a prometheus GaugeValue, but in the Readme for `windows_mssql_databases_data_files_size_bytes`, it's described as a counter. But it should be described as a gauge, right? Since the metric could go up or down? (And counters go up but also go up monotonically?)

Reply:
You're definitely right about `databases_data_files_size_bytes` - that's a gauge, and the docs are wrong.

The tricky part about interpreting the Microsoft docs on performance counters is that they actually describe the "formatted" data, which is what `perfmon` would display after applying some calculations. The "real" data that we get is before these calculations, which in some cases means we get total counts instead of averages, for example. I haven't seen them say that there are four different data points inside each one though, so I'm unsure what data you'll see here... 

Most of the time, you can look at the reported numbers and get a sense of what they represent, but if I'm not sure I look at the counter types. I've hacked together a small (and undocumented) tool for that [here](https://github.com/carlpett/perfcounter-inspector). Basically invoke it with the name of the counter (something like `MSSQL$SomeInstance:Wait Statistics` in this case) and it'll output all the fields with types. It probably makes sense here to check what types they have.

Reply:
OK, had to bang on that tool a bit to get it working in my dev env, but I did and got the following:

I first checked out "Locks" to use as a baseline (easy to find, has a variety of types):

<img width="701" alt="Screen Shot 2021-06-24 at 6 41 03 PM" src="https://user-images.githubusercontent.com/4651855/123345749-64179600-d51c-11eb-9ea5-7259a59fedfc.png">

So: AverageWaitTimems (perf_average_bulk), AveragetWaitTimeBase (perf_precision_timestamp) are all gauges, all the rest (perf_counter_bulk_count) are all counters -- except for LockWaitTimems. (That seems odd to me, but there's a general trend.)

So let's check out "Wait Statistics":

<img width="762" alt="Screen Shot 2021-06-24 at 6 41 10 PM" src="https://user-images.githubusercontent.com/4651855/123345758-67ab1d00-d51c-11eb-917c-2756bbcf64d9.png">

OK, they're all perf_counter_large_rawcount. 

For comparison, I checked "Access Methods" and all the perf_counter_large_rawcount entries (DeferredDroppedrowsets and DeferreddroppedAUs) are gauges, while the others are counters. 

BUT then I checked "Transactions" and there are several perf_counter_large_rawcount entries that are NOT gauges in the exporter (NonSnapshotVersionTransactions, SnapshotTransactions, UpdateSnapshotTransactions, etc.).

Is there some other way to determine whether a `perf_counter_large_rawcount` should be treated as a gauge or a counter?

(Also, updated the docs for `databases_data_files_size_bytes` since it was a minor fix.)

Reply:
Thanks for the analysis, and comparing with other collectors! From my understanding of the Microsoft documentation, `PERF_COUNTER_LARGE_RAWCOUNT` should map exactly to the Prometheus definition of a counter (_["A counter can wrap, as long as it does not wrap twice between snapshots"](https://docs.microsoft.com/en-us/previous-versions/windows/embedded/ms908616(v=msdn.10))_, and _["`PERF_COUNTER_LARGE_RAWCOUNT` [...] shows the instantaneous counter value"](https://docs.microsoft.com/en-us/previous-versions/windows/embedded/ms938549(v%3dmsdn.10))_).

So generally, I think we can assume that these should be counters, and that previous work in this collector can have had mistakes. Does the data make sense if interpreted as a counter? Does it match what `perfmon` says?

Reply:
OK, I changed all the Wait Statistics gauges to counters, checked `perfmon` for WaitStatistics, and things seemed adequately represented by the metrics exported that way. (That is, I see in perfmon that some stat goes up and down and I see the stat reflected when I refresh the localhost:9182/metrics page. 

(Unfortunately I don't have a great dev env for testing this particular sort of metrics, so a lot is 0, but maybe that's a feature: I can concentrate easily on the few that are moving...)

Reply:
Alright, let's say that this is a good point to merge and consider anything that doesn't line up a bug :) 
Many thanks for getting this worked through!

Reply:
Thanks for working through this with me!